Add new backlinks filter obs function. This is a helper function to w…

Happy0 opened the pull request:
…atch for links back to a message and add them to an observable list if they pass a given filter.

As discussed with @mmckegg on the %aJN7DTOIsTMbW/5x9yppL+PURCYFE+k2vzAl4E3Y4ww=.sha256 thread on scuttlebutt :D.

Happy0 commented:
I've tested with ssb-chess (my changes here: https://github.com/Happy0/ssb-chess/pull/43/files )

mmckegg commented on backlinks/obs_filter.js:
```diff
@@ -0,0 +1,80 @@
+  'sbot.pull.backlinks': 'first'
+})
+
+exports.gives = nest("backlinks.filter.obs", true)
```
The naming pattern is `topic.type.module`. As described here: [%G74TIna+iO1ByvYA79vNJyhd7xBiNeuDhLTmaWhBso0=.sha256](https://viewer.scuttlebot.io/%25G74TIna%2BiO1ByvYA79vNJyhd7xBiNeuDhLTmaWhBso0%3D.sha256)

So this should probably be `backlinks.obs.filter`.

mmckegg commented on backlinks/obs_filter.js:
```diff
@@ -0,0 +1,80 @@
+
+    var msgBacklinks = api.sbot.pull.backlinks({
+      query: [{
+        $filter: {
```
I see that you are passing in a function as a filter. But it is also interesting to note that the query here allows you to pass in JSON properties to filter. The advantage to the JSON filter instead of a function is that the filtering will happen in the sbot process which will make it a little bit more faster and efficient (since it doesn't have to send all the extra data down the wire).

The filter method makes sense, but an idea for improvement and making this module more versatile (maybe later) would be to also accept an object as a filter, and then inject it into the query filter.
 
It is using this: https://github.com/dominictarr/map-filter-reduce

mmckegg commented on backlinks/obs_filter.js:
```diff
@@ -0,0 +1,80 @@
+    var msgFilter = pull.filter((msg) => {
+      // We do not include the 'sync' message which indicates that any
+      // further messages are newly gossiped messages in the list of messages.
+      if (msg.sync) {
```
`mutant-pull-reduce` [actually already handles the sync message](https://github.com/mmckegg/mutant-pull-reduce/blob/master/index.js#L39), and it exposes an `obs.sync`. So this isn't really needed.

At the moment, it defaults to doing this, but it should really be an option. I suggest that if you use this, you should add `{sync: true}` to the `MutantPullReduce` options.

Happy0 replied:
Huh. That's weird. I originally added this code because in ssb-chess I already have a watch on obs.sync and it wasn't getting fired (therefore the games weren't getting rendered), so I assumed I had to take care of the sync myself.

I just removed this stuff, and then set a breakpoint on https://github.com/mmckegg/mutant-pull-reduce/blob/master/index.js#L39 and I notice that it's never getting set to true (i.e. that line is never executed.) To make sure I was breakpointing the right file, I added a breakpoint to the line that pushes the item and that did work.

I was wondering if you could help me debug this?

I'll push my updates to this pull request (including setting the 'sync' observable manually) for now and we can try and fix the bug in mutant-pull-reduce before merging this.

Happy0 commented:
I changed the 'gives' name to meet the convention you mentioned.

I've also removed the 'sync' observable I was setting manually. However, I think there might be a bug in mutant-pull-reduce as its 'sync' value is never being set as I mentioned here: https://github.com/ssbc/patchcore/pull/46#discussion_r153084050 . We should investigate this before merging.

I haven't dealt with the 'filter by object directly in an sbot query' for now. However, I've made it easier to expand the function to have an option to do that in the future by accepting `opts` as a second parameter. In the future, this could accept an optional filter (a function) or an optional object of fields to filter by in an sbot query. Let me know if this approach could be improved.

Happy0 commented:
Oh wait, simple explanation. I was using `live:true` rather than `sync:true` for my argument for `sbot.pull.backlinks`.

I guess this is ready to review again :).

You can use my ssb-chess Pull Request to help you test, if you like :)

Happy0 commented:
@mmckegg just wanted to check if you want the optional 'filter by fields in sbot query' is a prerequisite for this getting merged?

mmckegg commented:
@Happy0 

No, this is looking pretty good! Will do a little testing then hopefully merge this later today 😃

mmckegg commented:
Published as **v1.20.0** 🎉

Happy0 commented:
\o/ :D
